source/common/router/router.cc

```diff
@@ -423,20 +442,27 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
     do_shadowing_ = false;
   }
 
-  if (buffering) {
-    // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
-    // since it's all moves from here on.
-    Buffer::OwnedImpl copy(data);
-    upstream_request_->encodeData(copy, end_stream);
+  for (unsigned long i = 0; i < upstream_requests_.size(); i++) {
+    if (buffering) {
+      // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
+      // since it's all moves from here on.
+      Buffer::OwnedImpl copy(data);
+      upstream_requests_[i]->encodeData(copy, end_stream);
+
+      if (i == 0) {
+      }
+    } else {
+      upstream_requests_[i]->encodeData(data, end_stream);
+    }
+  }
 
+  if (buffering) {
     // If we are potentially going to retry or shadow this request we need to buffer.
     // This will not cause the connection manager to 413 because before we hit the
     // buffer limit we give up on retries and buffering. We must buffer using addDecodedData()
     // so that all buffered data is available by the time we do request complete processing and
     // potentially shadow.
     callbacks_->addDecodedData(data, true);
```

Comment:
I'm assuming we only want to call this once rather than per upstream request

Reply:
Yes, definitely.